packages/realm/src/app-services/SyncSession.ts

```diff
@@ -229,106 +228,90 @@ const mockApp = new Proxy({} as App, {
 const PROGRESS_LISTENERS = new Listeners<
   ProgressNotificationCallback,
   ListenerToken,
-  [binding.SyncSession, ProgressDirection, ProgressMode]
+  [binding.WeakSyncSession, binding.SyncSession, ProgressDirection, ProgressMode]
 >({
   throwOnReAdd: true,
-  add(callback, internal, direction, mode) {
+  add(callback, weakInternal, internal, direction, mode) {
     const token = internal.registerProgressNotifier(
       (transferred, transferable) => callback(Number(transferred), Number(transferable)),
       toBindingDirection(direction),
       mode === ProgressMode.ReportIndefinitely,
     );
-    return { internal, token };
+    return { weakInternal, token };
   },
-  remove({ internal, token }) {
-    return internal.unregisterProgressNotifier(token);
+  remove({ weakInternal, token }) {
+    weakInternal.withDeref((internal) => internal?.unregisterProgressNotifier(token));
   },
 });
 
 /**
  * Connection listeners are shared across instances of the SyncSession, making it possible to deregister a listener on another session
  * TODO: Consider adding a check to verify that the callback is removed from the correct SyncSession (although that would break the API)
  */
-const CONNECTION_LISTENERS = new Listeners<ConnectionNotificationCallback, ListenerToken, [binding.SyncSession]>({
+const CONNECTION_LISTENERS = new Listeners<
+  ConnectionNotificationCallback,
+  ListenerToken,
+  [binding.WeakSyncSession, binding.SyncSession]
+>({
   throwOnReAdd: true,
-  add(callback, internal) {
+  add(callback, weakInternal, internal) {
     const token = internal.registerConnectionChangeCallback((oldState, newState) =>
       callback(fromBindingConnectionState(newState), fromBindingConnectionState(oldState)),
     );
-    return { internal, token };
+    return { weakInternal, token };
   },
-  remove({ internal, token }) {
-    internal.unregisterConnectionChangeCallback(token);
+  remove({ weakInternal, token }) {
+    weakInternal.withDeref((internal) => internal?.unregisterConnectionChangeCallback(token));
   },
 });
 
 export class SyncSession {
-  private static instances = new Set<binding.WeakRef<SyncSession>>();
```

Comment:
I'm pretty sure we can kill this off now that we are using `weak_ptr` since a long-lived `SyncSession` won't keep the underlying C++ object alive anymore.

Reply:
That would be great 🙏